rate_limits.go

```diff
@@ -0,0 +1,80 @@
+package stream_chat // nolint: golint
+
+import (
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// RateLimitInfo represents the quota and usage for a single endpoint.
+type RateLimitInfo struct {
+	// Limit is the maximum number of API calls for a single time window (1 minute).
+	Limit int64 `json:"limit"`
+	// Remaining is the number of API calls remaining in the current time window (1 minute).
+	Remaining int64 `json:"remaining"`
+	// Reset is the Unix timestamp of the expiration of the current rate limit time window.
+	Reset int64 `json:"reset"`
+}
+
+// RateLimitsMap holds the rate limit information, where the keys are the names of the endpoints and the values are
+// the related RateLimitInfo containing the quota, usage, and reset data.
+type RateLimitsMap map[string]RateLimitInfo
+
+// ResetTime is a simple helper to get the time.Time representation of the Reset field of the given limit window.
+func (i RateLimitInfo) ResetTime() time.Time {
+	return time.Unix(i.Reset, 0)
+}
+
+// GetRateLimitsResponse is the response of the Client.GetRateLimits call. It includes, if present, the rate
+// limits for the supported platforms, namely server-side, Android, iOS, and web.
+type GetRateLimitsResponse struct {
+	ServerSide RateLimitsMap `json:"server_side,omitempty"`
+	Android    RateLimitsMap `json:"android,omitempty"`
+	IOS        RateLimitsMap `json:"ios,omitempty"`
+	Web        RateLimitsMap `json:"web,omitempty"`
+}
+
+// GetRateLimitsOptions configures the Client.GetRateLimits call.
+type GetRateLimitsOptions struct {
+	// ServerSide, if true, restricts the returned limits to server-side clients only.
+	ServerSide bool
+	// Android, if true, restricts the returned limits to Android clients only.
+	Android bool
+	// IOS, if true, restricts the returned limits to iOS clients only.
+	IOS bool
+	// Web, if true, restricts the returned limits to web clients only.
+	Web bool
+	// Endpoints restricts the returned limits info to the specified endpoints.
+	Endpoints []string
+}
+
+// GetRateLimits returns the current rate limit quotas and usage. If no params are toggled, all the limits
+// for all platforms are returned.
+func (c *Client) GetRateLimits(options GetRateLimitsOptions) (GetRateLimitsResponse, error) {
```

Comment:
That is fine but in general, having a struct here is making harder to change it later.

Having `...options` is better from API perspective.

`GetRateLimits(Android, IOS, Endpoints("query_channels", "query_users")) ..`

Reply:
I agree on paper with the variadic options, but I think in this case it would be better to keep it as a single struct. The alternative is to have e.g.

```go
type RateLimitParams struct {
...
}

func RateLimitOptions func(p *RateLimitParams)

func GetRateLimits(opts ...RateLimitOptions) {
	// ...
	params := RateLimitParams{}
	for _, opt := range options {
		opt(&params)
	}
	// ...
```

which imho doesn't add much value

Reply:
happy to have the change above ^ if and when there are non-default values in the params

Reply:
Having params non-exported and variadic options enables us to change anything without breaking as we want even if we can't see how from today.
 
```go
type rateLimitParams struct {
...
}
```

However, it's good for now.

Reply:
@ferhatelmas updated